report-parser/src/main.rs

```diff
@@ -561,6 +658,162 @@ fn parse_bundler_audit_plaintext(
     Ok(events)
 }
 
+#[derive(Deserialize, Debug)]
+struct PythonSafety {
+    affected_package: String,
+    affected_versions: String,
+    installed_version: String,
+    advisory_description: String,
+    advisory_id: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct MetaInfo {
+    advisory: String,
+    timestamp: Value,
+}
+
+#[derive(Deserialize, Debug)]
+struct SafetyPackageVulnInfo {
+    advisory: String,
+    cve: Value,
+    id: String,
+    specs: Vec<String>,
+    v: String,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(untagged)]
+enum SafetyJsonData {
+    Vuln(Vec<SafetyPackageVulnInfo>),
+    Meta(MetaInfo),
+}
+
+fn download_and_parse_python_safety_vulns(
+    server_name: String,
+    etag: &mut Option<String>,
+    client: &Client,
+) -> Result<Option<HashMap<String, Option<String>>>, Box<dyn Error>> {
+    let safety_json_db_url = &server_name.to_string();
+    let head_resp = client.head(safety_json_db_url).send()?;
+    if head_resp.status().is_success() {
+        if let Some(etag_new) = head_resp.headers().get(ETAG) {
+            // If the etag passed in is none or different to the one we just got, then download below....
+            match etag {
+                Some(etag_old) => {
+                    if *etag_old == etag_new.to_str().unwrap() {
+                        return Ok(None);
+                    } else {
+                        *etag = Some(etag_new.to_str().unwrap().to_owned());
+                    }
+                }
+                None => *etag = Some(etag_new.to_str().unwrap().to_owned()),
+            }
+        }
+    } else {
+        return Err(Box::new(
+            err_msg(format!(
+                "Unable to grab head from python safety database: ({})",
+                head_resp.status()
+            ))
+            .compat(),
+        )
+        .into());
+    }
+
+    let meta_resp_text = reqwest::blocking::get(safety_json_db_url)?.text()?;
+    let python_safety_vuln_info_json: HashMap<String, SafetyJsonData> =
+        serde_json::from_str(meta_resp_text.as_ref())?;
+    let cve_items = python_safety_vuln_info_json
+        .values()
+        .filter(|ref _s| match _s {
+            SafetyJsonData::Vuln(_s) => true,
+            _ => false,
+        })
+        .map(|s| match s {
+            SafetyJsonData::Vuln(s) => s.iter(),
+            _ => unreachable!(),
+        })
+        .collect::<Vec<_>>()
+        .into_iter()
+        .flatten()
+        .collect::<Vec<_>>()
+        .iter()
```

Comment:
```suggestion
        .flatten()
```
Types before:
![image](https://user-images.githubusercontent.com/15730032/105709289-6940cc80-5f0d-11eb-9119-e557bb500b44.png)
Types after:
![image](https://user-images.githubusercontent.com/15730032/105709304-6d6cea00-5f0d-11eb-87a3-00f26e5d8a1f.png)